### ISSUE TITLE
CORS issue

### ISSUE THREAD

**Issue by chongma** [dsb-norge/vue-keycloak-js#21]
I am using Keycloak 4.5.0 and I seem to bump into a CORS issue.

`Access to XMLHttpRequest at 'https://securetest.mydomain.com/auth/realms/homeconvey-production/protocol/openid-connect/token' from origin 'http://localhost:8081' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.`

I have added the Web Origin 'http://localhost:8081' to the Client that I am using.

**Comment by chongma** [comment]
Is it possibly related to not being able to pass a client secret?  When I pass a clientSecret in options it does not show in the console.log of the keycloak object.  Is there any way of just passing a keycloak.json directly to the Keycloak()?

**Comment by anderius** [comment]
Client secret is not related to CORS, and would not make sense with the javascript adapter. The secret would have to be revealed _before_ authentication, and would no longer be secret.

Your Keycloak instance should add Access-Control-Allow-* headers. Maybe you have a proxy in front that removes them?

You can try with cURL, please notice the Origin header (the browser adds this normally):

`curl -X OPTIONS https://securetest.mydomain.com/auth/realms/homeconvey-production/protocol/openid-connect/token -H 'Origin: http://localhost'`

That should give you response headers like this:

`Access-Control-Allow-Headers →Origin, Accept, X-Requested-With, Content-Type, Access-Control-Request-Method, Access-Control-Request-Headers, Authorization`
`Access-Control-Allow-Methods →POST, OPTIONS`
`Access-Control-Allow-Origin →http://localhost`
`Access-Control-Allow-Credentials →true`

**Comment by chongma** [comment]
Yes I am behind an apache httpd proxy
There are no response headers when issuing the curl command.  I also issued behind the proxy on the server with
`curl -X OPTIONS http://localhost:8380/auth/realms/homeconvey-production/protocol/openid-connect/token -H 'Origin: http://localhost'`
It seems like configuring JBoss for https proxy has somehow stripped the access control headers.  I will investigate and report back here

**Comment by chongma** [comment]
The original command didn't work.  I added verbose output
`curl -vX OPTIONS https://securetest.mydomain.com/auth/realms/homeconvey-production/protocol/openid-connect/token -H 'Origin: http://localhost'`
and it does seem to be working correctly
`< HTTP/1.1 200 OK`
`< Date: Wed, 15 May 2019 09:36:56 GMT`
`< Server: Apache`
`< Access-Control-Allow-Origin: http://localhost`
`< Access-Control-Allow-Headers: Origin, Accept, X-Requested-With, Content-Type, Access-Control-Request-Method, Access-Control-Request-Headers, Authorization`
`< Access-Control-Allow-Credentials: true`
`< Content-Length: 0`
`< Access-Control-Allow-Methods: POST, OPTIONS`
`< Access-Control-Max-Age: 3600`

**Comment by chongma** [comment]
ok, it works if i remove secret from the client.  but how come keycloak.js accepts a clientSecret if it is not allowed to be used?  it would add a tiny layer of security if the attacker knew the location of keycloak server but not of client webapp?

**Comment by anderius** [comment]
clientSecret is not documented as far as I can see, and I have no idea why the Keycloak team included handling of it in the source code.

You already have that kind of "security", because the clientId is as secret as the clientSecret in a javascript application.

I am closing this issue now, as it is not directly related to the Vue wrapper, butrelates to Keycloak javascript adapter and the Keycloak server itself.

**Comment by anderius** [comment]
I have to modify myself, as client secret _could_ be related to CORS. The presence of the client secret make the keycloak javascript adapter use an Authorization header with Basic authentication. That header is, however, on the list of allowed headers, and should work.

**Comment by chongma** [comment]
Is there any way to pass clientSecret through to test the behaviour?

**Comment by nlien** [comment]
If you want to pass it on, you can, as far as I can read the sourcecode for the Keycloak adapter, pass it on the following way:

```
Vue.use(VueKeyCloak, {
  config: {
    'realm': 'xxxx',
    'url': 'xxxx',
    'clientId': 'xxxx',
    'credentials': {
      'secret': 'xxxx' // clientSecret
  },
  onReady: (keycloak) => {
    new Vue({
      el: '#app',
      router,
      template: '<App/>',
      render: h => h(App)
    })
  }
})
```

**Comment by chongma** [comment]
@nlien yep that works perfectly.  thanks

**Comment by PeterPilley** [comment]
Hi Sorry to re-open this, but we are experiencing this with Vue3.

it only happens with our vue deployments the same in keycloak-connect with nodejs, or with python are fine. I tried the curl command to test the response and it report that it is allowed.

Interestingly if I access the link in a seperate browser and then come back to our vue deployment it works, does this mean that something is not being set in the session correctly?

I have tried setting the web origins and redirects in keycloak without any change.
